src/main/java/fi/okm/jod/yksilo/service/KoskiService.java

```diff
@@ -0,0 +1,138 @@
+/*
+ * Copyright (c) 2025 The Finnish Ministry of Education and Culture, The Finnish
+ * The Ministry of Economic Affairs and Employment, The Finnish National Agency of
+ * Education (Opetushallitus) and The Finnish Development and Administration centre
+ * for ELY Centres and TE Offices (KEHA).
+ *
+ * Licensed under the EUPL-1.2-or-later.
+ */
+
+package fi.okm.jod.yksilo.service;
+
+import com.jayway.jsonpath.JsonPath;
+import fi.okm.jod.yksilo.domain.Kieli;
+import fi.okm.jod.yksilo.domain.LocalizedString;
+import fi.okm.jod.yksilo.dto.profiili.KoulutusDto;
+import java.net.URI;
+import java.time.LocalDate;
+import java.util.Collections;
+import java.util.EnumMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.http.HttpHeaders;
+import org.springframework.http.converter.json.MappingJackson2HttpMessageConverter;
+import org.springframework.stereotype.Service;
+import org.springframework.web.client.RestClient;
+
+@Service
+@Slf4j
+public class KoskiService {
+  public static final Set<String> ALLOWED_HOSTS =
+      Set.of("https://opintopolku.fi", "https://testiopintopolku.fi");
+
+  private final RestClient restClient;
+
+  public KoskiService(
+      RestClient.Builder restClientBuilder,
+      MappingJackson2HttpMessageConverter messageConverter,
+      @Value("") String baseUrl) {
+    log.info("Creating KoskiService, baseUrl: {}", baseUrl);
+
+    this.restClient =
+        restClientBuilder
+            .messageConverters(
+                converters -> {
+                  converters.clear();
+                  converters.add(messageConverter);
+                })
+            .baseUrl(baseUrl)
+            .defaultHeader(
+                HttpHeaders.USER_AGENT, "fi.okm.jod (https://okm.fi/hanke?tunnus=OKM069:00/2021)")
+            .build();
+  }
+
+  private <T> T readJsonProperty(Object json, String jsonPathExpression) {
+    try {
+      return JsonPath.read(json, jsonPathExpression);
+    } catch (Exception e) {
+      log.error("JSON expression error", e);
+      return null;
+    }
+  }
+
+  public List<KoulutusDto> getKoskiData(URI linkki) {
+    var h = linkki.getScheme() + "://" + linkki.getHost();
+    var location = ALLOWED_HOSTS.stream().filter(a -> a.equals(h)).findFirst();
```

Comment:
`if (!ALLOWED_HOSTS.contains(linkki.getHost())) { throw ...}` would be simpler (and more efficient).